clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java

```diff
@@ -56,6 +59,8 @@
     public ProducerRecord(String topic, Integer partition, Long timestamp, K key, V value) {
         if (topic == null)
             throw new IllegalArgumentException("Topic cannot be null");
+        if (!TopicValidator.isValidateTopicName(topic))
+            throw new InvalidTopicException(topic + " is not a valid topic name.");
```

Comment:
My suggestion would be to move the logic from the core module `kafka.common.Topic.scala` to  the client module. And then replace all usages of the old Topic class in core with the new version added in clients.